Implement (if-let) guards in pattern matchings

maximebuyse opened the pull request:
Closes #814

This adds support for guards in pattern matchings. They are added to the AST (with a corresponding feature guard), and handled in the THIR import. A phase rewrites them as an equivalent nested pattern matching. Apart from supporting guards as specified in #807, this will also be useful for future phases that will generate guards (see #806).

Example of code with if-let guard (which is currently a nightly feature) and Rust code for the equivalent after the new phase:
```
pub fn if_let_guard(x: Option<Result<i32, i32>>) -> i32 {
    match x {
        None => 0,
        Some(v) if let Ok(y) = v => y,
        Some(Err(y)) => y,
        _ => 1,
    }
}

pub fn equivalent(x: Option<Result<i32, i32>>) -> i32 {
    match x {
        None => 0,
        _ => match match x {
            Some(v) => match v {
                Ok(y) => Some(y),
                _ => None,
            },
            _ => None,
        } {
            Some(y) => y,
            None => match x {
                Some(Err(y)) => y,
                _ => 1,
            },
        },
    }
}
```
Regular if guards are also supported (but for now desugared as if-let guards):
```
pub fn if_guard(x: Option<i32>) -> i32 {
    match x {
        Some(v) if v > 0 => v,
        _ => 0,
    }
}

pub fn equivalent(x: Option<i32>) -> i32 {
    match x {
        Some(v) if let true = v > 0 => v,
        _ => 0,
    }
}
```

W95Psp commented on engine/lib/ast.ml:
```diff
@@ -338,8 +338,11 @@ functor
           witness : F.nontrivial_lhs;
         }
 
+    and guard = { guard : guard'; span : span }
```
Can you add:
 - a doc comment that explains what is a `guard` (with an example)
 - a comment on `guard'` explaining why there is only one variant currently, and a link to an issue that explains the roadmap for the work left on guards (that is, simple lets and let-chains), with link to Rust documentation on that

W95Psp commented on engine/lib/ast_utils.ml:
```diff
@@ -410,8 +410,12 @@ module Make (F : Features.T) = struct
-          method! visit_arm' env { arm_pat; body } =
-            shadows ~env [ arm_pat ] body super#visit_expr
+          method! visit_arm' env { arm_pat; body; guard } =
+            match guard with
```
Here, you can remove duplicate code with a `shadows ~env [ arm_pat ] rhs super#visit_expr ++ (Option.map ~f:(...) guard |> Option.value ~default_or_thunk:self#zero))`

maximebuyse replied:
I think it is not possible here as in the `None` branch the expression on which we call `shadows` is `body`. But maybe another simplification is possible

W95Psp replied:
Ah, right, that makes sense!

W95Psp commented on engine/lib/ast_utils.ml:
```diff
@@ -601,8 +607,17 @@ module Make (F : Features.T) = struct
-        method! visit_arm' () { arm_pat; body } =
-          without_pat_vars (self#visit_expr () body) arm_pat
+        method! visit_arm' () { arm_pat; body; guard } =
+          match guard with
+          | Some { guard = IfLet { lhs; rhs; _ }; _ } ->
+              let rhs_vars =
+                without_pat_vars (self#visit_expr () rhs) arm_pat
+              in
+              let body_vars =
+                without_pats_vars (self#visit_expr () body) [ arm_pat; lhs ]
+              in
+              Set.union rhs_vars body_vars
+          | None -> without_pat_vars (self#visit_expr () body) arm_pat
```
You can use a similar strategy as above, with an `Option.map`

W95Psp commented on engine/lib/ast_utils.ml:
```diff
@@ -777,6 +792,9 @@ module Make (F : Features.T) = struct
 
   let make_wild_pat (typ : ty) (span : span) : pat = { p = PWild; span; typ }
 
+  let make_unguarded_arm (arm_pat : pat) (body : expr) (span : span) : arm =
+    { arm = { arm_pat; body; guard = None }; span }
```
Can you make this into a `make_arm`, with an optional (defaulting to `None`) `?guard` argument?

W95Psp commented on engine/lib/generic_printer/generic_printer.ml:
```diff
@@ -430,10 +430,17 @@ module Make (F : Features.T) (View : Concrete_ident.VIEW_API) = struct
             let body = print#expr_at Arm_body body in
-            pat ^^ string " => " ^^ body ^^ comma
+            let g =
+              match guard with
+              | Some { guard = IfLet { lhs; rhs; _ }; _ } ->
+                  string " if let " ^^ print#pat_at Arm_pat lhs ^^ string " = "
+                  ^^ print#expr_at Arm_body rhs
+              | None -> string ""
```
Can you:
 - use `empty` instead of `string ""`
 - use a `Option.map ~f:(...) guard |> Option.value ~default:empty`?

W95Psp replied:
Oh and why not renaming `g` into `guard` here?

W95Psp commented on engine/lib/import_thir.ml:
```diff
@@ -418,7 +418,10 @@ end) : EXPR = struct
                   arms =
                     [
-                      { arm = { arm_pat = lhs; body }; span = lhs_body_span };
+                      {
+                        arm = { arm_pat = lhs; body; guard = None };
+                        span = lhs_body_span;
+                      };
```
In this file we should probably use `make_(unguarded_)arm` now :)

W95Psp commented on engine/lib/import_thir.ml:
```diff
@@ -1091,7 +1095,36 @@ end) : EXPR = struct
+    let guard =
+      Option.map
+        ~f:(fun (e : Thir.decorated_for__expr_kind) ->
+          match e.contents with
+          | Let { expr; pat } ->
+              {
+                guard =
+                  IfLet
+                    {
+                      lhs = c_pat pat;
+                      rhs = c_expr expr;
+                      witness = W.match_guard;
+                    };
+                span = Span.of_thir e.span;
+              }
+          | _ ->
+              {
+                guard =
+                  IfLet
+                    {
+                      lhs =
+                        { p = PConstant { lit = Bool true }; span; typ = TBool };
+                      rhs = c_expr e;
+                      witness = W.match_guard;
+                    };
+                span = Span.of_thir e.span;
+              })
+        arm.guard
```
Can you put the match in a `let` binding? e.g. a `let guard = match ... with ... in {guard; span = Span.of_thir ...}`?

W95Psp commented on engine/lib/phases/phase_drop_match_guards.ml:
```diff
@@ -0,0 +1,200 @@
+      | [] -> treated
+      | { arm = { arm_pat; body; guard = None }; span } :: remaining ->
+          let new_arm : B.arm =
+            {
+              arm = { arm_pat = dpat arm_pat; body = dexpr body; guard = None };
+              span;
+            }
```
Maybe use `U.make_arm`?

maximebuyse commented:
This also closes #818 and #807

W95Psp commented on engine/lib/phases/phase_drop_match_guards.ml:
```diff
@@ -0,0 +1,200 @@
+                        UB.make_unguarded_arm (dpat arm_pat)
+                          {
+                            e =
+                              Match
```
Can you hoist this inner match to a let binding? I mean in OCaml, not in the generated code.
Then, can you add a comment that explains what is this sub-match exactly, with an example (e.g. a quote of the example you'll have earlier in the code)

W95Psp commented on engine/lib/phases/phase_drop_match_guards.ml:
```diff
@@ -0,0 +1,200 @@
+          span;
+        }
+        :: remaining ->
+          let result_typ = dty span body.typ in
```
Here, can you add a comment with an example in Rust syntax (you can use `<lhs>`, `<rhs>`, `<body>` et `<arm_pat>` in the example)?
I.e. an example of the matched arm + what will be the rewrite?

W95Psp commented on engine/lib/phases/phase_drop_match_guards.ml:
```diff
@@ -0,0 +1,200 @@
+      match expr with
+      | [%inline_arms "dexpr'.*" - Match] -> auto
+      | Match { scrutinee; arms } ->
+          let arms = transform_arms (dexpr scrutinee) (List.rev arms) [] in
```
If you have `match <scrut> { ... if ... => ..., <other arms> }`, this would generate `match <scrut> { _ => match <scrut>} {...}}`, we should optimize that

maximebuyse replied:
Yes, I also noticed that with tests. This phase can generate a lot of nested pattern matchings, some of them with only one Wildcard branch or other simplifiable situations. Do you think we should optimize in this phase (trying to simpler code in specific situations) or in another one (that would catch them and simplify)?

W95Psp replied:
I think we should do that in this phase: if a user writes a match like that in Rust, it's better not to rewrite it.
We should collapse matches only when a rewrite was done I think

W95Psp replied:
Most of the unnecessary `match ... { _ => ... }` seems to be gone :smiley: but if one appears at top-level, it seems like it's still there, e.g. https://hax-playground.cryspen.com/#fstar/9dc02d2817/gist=f490560c1ca7bb7a0e30f0e133fd8202
I guess you need to do something right there, when calling `transform_arms` in `dexpr'`, no?

maximebuyse replied:
Good catch! Should be fixed now.

W95Psp commented on engine/lib/phases/phase_drop_match_guards.ml:
```diff
@@ -0,0 +1,200 @@
+          in
+
+          let expr_none = mk_opt_expr None in
+          let r : B.expr =
```
Can you add a comment here, with a example, explaining what AST we're building?

W95Psp commented on engine/lib/phases/phase_drop_match_guards.ml:
```diff
@@ -0,0 +1,200 @@
+(* This phase removes guards from pattern matchings. It rewrites
```
We should include a succinct rewrite example here.

W95Psp commented on engine/lib/print_rust.ml:
```diff
@@ -278,8 +278,14 @@ module Raw = struct
-            ~f:(fun { arm = { arm_pat; body }; _ } ->
-              ppat arm_pat & !" => {" & pexpr body & !"}")
+            ~f:(fun { arm = { arm_pat; body; guard }; _ } ->
+              let g =
+                match guard with
+                | Some { guard = IfLet { lhs; rhs; _ }; _ } ->
+                    !" if let " & ppat lhs & !" = " & pexpr rhs
+                | None -> !""
+              in
```
Similar comments that on the generic printer: maybe use `Option.map`&`Option.value` and name `g` `guard`?

W95Psp reviewed (changes requested):
Thank you, the output looks like what we want!
However, we need:
 - more tests (e.g. a match where more than one arm as an `if`/`if let`, exercise `if`, exercise a non-trivial `if let` with a computation in the RHS... you don't need to have the equivalent desugar for everything)
 - some description on the PR. Your PR should:
    + explain the problem you solve (referring to an issue is fine, iff the issue has a nice and intelligible description about the exact problem we are solving, with examples --but in your case the issue has no description, please add some as requested in the issue)
    + then give an overview of technically what is our approach
    + then give examples of what one can do if your PR is merged (possibly with links to playground)
    + note that those points above should basically be copy pastes from code documentations or tests cases

maximebuyse commented:
> Thank you, the output looks like what we want! However, we need:
> 
>     * more tests (e.g. a match where more than one arm as an `if`/`if let`, exercise `if`, exercise a non-trivial `if let` with a computation in the RHS... you don't need to have the equivalent desugar for everything)
> 
>     * some description on the PR. Your PR should:
>       
>       * explain the problem you solve (referring to an issue is fine, iff the issue has a nice and intelligible description about the exact problem we are solving, with examples --but in your case the issue has no description, please add some as requested in the issue)
>       * then give an overview of technically what is our approach
>       * then give examples of what one can do if your PR is merged (possibly with links to playground)
>       * note that those points above should basically be copy pastes from code documentations or tests cases

Thank you. I addressed most of your feedback but I can add more tests / documentation if you think it is not clear enough.

maximebuyse commented:
> Thanks you, that's great! :) I don't know how you want to do this optimization for "one wildcard arm" matches: in this PR or in a subsequent one?

I'll do it in this one

maximebuyse commented:
> Thanks you, that's great! :) I don't know how you want to do this optimization for "one wildcard arm" matches: in this PR or in a subsequent one?

@W95Psp That's done now

W95Psp commented on engine/lib/phases/phase_drop_match_guards.ml:
```diff
@@ -60,12 +60,19 @@ module%inlined_contents Make (F : Features.T) = struct
       | Match { scrutinee; arms } ->
           let arms = transform_arms (dexpr scrutinee) (List.rev arms) [] in
-          Match { scrutinee = dexpr scrutinee; arms }
+          maybe_simplified_match (dexpr scrutinee) arms
```
Sorry for being picky :grin: but can do `maybe_simplified_match` iff the user did not type a match with one arm with `_` LHS and no guard? otherwise such user typed matches are optimized away, e.g. https://hax-playground.cryspen.com/#fstar/365ad6f8bf/gist=be0c4a2ebe27079803f981fc39c45283

maximebuyse replied:
No problem, I fixed that now!

W95Psp commented:
Thanks, I think that's great now! Let's merge that PR :)
